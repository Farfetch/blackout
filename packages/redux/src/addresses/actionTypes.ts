/**
 * Action type dispatched when the fetch prediction request fails.
 */
export const FETCH_ADDRESS_PREDICTIONS_FAILURE =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_FAILURE';
/**
 * Action type dispatched when the fetch prediction request starts.
 */
export const FETCH_ADDRESS_PREDICTIONS_REQUEST =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_REQUEST';
/**
 * Action type dispatched when the fetch prediction request succeeds.
 */
export const FETCH_ADDRESS_PREDICTIONS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTIONS_SUCCESS';

/**
 * Action type dispatched when the fetch prediction details request fails.
 */
export const FETCH_ADDRESS_PREDICTION_FAILURE =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_FAILURE';
/**
 * Action type dispatched when the fetch prediction details request starts.
 */
export const FETCH_ADDRESS_PREDICTION_REQUEST =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_REQUEST';
/**
 * Action type dispatched when the fetch prediction details request succeeds.
 */
export const FETCH_ADDRESS_PREDICTION_SUCCESS =
  '@farfetch/blackout-redux/FETCH_ADDRESS_PREDICTION_SUCCESS';

/**
 * Action type dispatched when resetting the address prediction.
 */
export const RESET_ADDRESS_PREDICTIONS =
  '@farfetch/blackout-redux/RESET_ADDRESS_PREDICTIONS';
