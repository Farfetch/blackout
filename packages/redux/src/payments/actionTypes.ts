/**
 * Action type dispatched when the create payment intent instrument request fails.
 */
export const CREATE_PAYMENT_INTENT_INSTRUMENT_FAILURE =
  '@farfetch/blackout-redux/CREATE_PAYMENT_INTENT_INSTRUMENT_FAILURE';
/**
 * Action type dispatched when the create payment intent instrument request starts.
 */
export const CREATE_PAYMENT_INTENT_INSTRUMENT_REQUEST =
  '@farfetch/blackout-redux/CREATE_PAYMENT_INTENT_INSTRUMENT_REQUEST';
/**
 * Action type dispatched when the create payment intent instrument request succeeds.
 */
export const CREATE_PAYMENT_INTENT_INSTRUMENT_SUCCESS =
  '@farfetch/blackout-redux/CREATE_PAYMENT_INTENT_INSTRUMENT_SUCCESS';

/**
 * Action type dispatched when the fetch payment intent instrument request fails.
 */
export const FETCH_PAYMENT_INTENT_INSTRUMENT_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_INSTRUMENT_FAILURE';
/**
 * Action type dispatched when the fetch payment intent instrument request starts.
 */
export const FETCH_PAYMENT_INTENT_INSTRUMENT_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_INSTRUMENT_REQUEST';
/**
 * Action type dispatched when the fetch payment intent instrument request succeeds.
 */
export const FETCH_PAYMENT_INTENT_INSTRUMENT_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_INSTRUMENT_SUCCESS';

/**
 * Action type dispatched when the update payment intent instrument request fails.
 */
export const UPDATE_PAYMENT_INTENT_INSTRUMENT_FAILURE =
  '@farfetch/blackout-redux/UPDATE_PAYMENT_INTENT_INSTRUMENT_FAILURE';
/**
 * Action type dispatched when the update payment intent instrument request starts.
 */
export const UPDATE_PAYMENT_INTENT_INSTRUMENT_REQUEST =
  '@farfetch/blackout-redux/UPDATE_PAYMENT_INTENT_INSTRUMENT_REQUEST';
/**
 * Action type dispatched when the update payment intent instrument request succeeds.
 */
export const UPDATE_PAYMENT_INTENT_INSTRUMENT_SUCCESS =
  '@farfetch/blackout-redux/UPDATE_PAYMENT_INTENT_INSTRUMENT_SUCCESS';

/**
 * Action type dispatched when the remove payment intent instrument request fails.
 */
export const REMOVE_PAYMENT_INTENT_INSTRUMENT_FAILURE =
  '@farfetch/blackout-redux/REMOVE_PAYMENT_INTENT_INSTRUMENT_FAILURE';
/**
 * Action type dispatched when the remove payment intent instrument request starts.
 */
export const REMOVE_PAYMENT_INTENT_INSTRUMENT_REQUEST =
  '@farfetch/blackout-redux/REMOVE_PAYMENT_INTENT_INSTRUMENT_REQUEST';
/**
 * Action type dispatched when the remove payment intent instrument request succeeds.
 */
export const REMOVE_PAYMENT_INTENT_INSTRUMENT_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_PAYMENT_INTENT_INSTRUMENT_SUCCESS';

/**
 * Action type dispatched when the fetch payment intent instruments request fails.
 */
export const FETCH_PAYMENT_INTENT_INSTRUMENTS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_INSTRUMENTS_FAILURE';
/**
 * Action type dispatched when the fetch payment intent instruments request starts.
 */
export const FETCH_PAYMENT_INTENT_INSTRUMENTS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_INSTRUMENTS_REQUEST';
/**
 * Action type dispatched when the fetch payment intent instruments request succeeds.
 */
export const FETCH_PAYMENT_INTENT_INSTRUMENTS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_INSTRUMENTS_SUCCESS';

/**
 * Action type dispatched when resetting the payment instruments.
 */
export const RESET_PAYMENT_INSTRUMENTS_STATE =
  '@farfetch/blackout-redux/RESET_PAYMENT_INSTRUMENTS_STATE';

/**
 * Action type dispatched when the remove payment token request fails.
 */
export const REMOVE_PAYMENT_TOKEN_FAILURE =
  '@farfetch/blackout-redux/REMOVE_PAYMENT_TOKEN_FAILURE';
/**
 * Action type dispatched when the remove payment token request starts.
 */
export const REMOVE_PAYMENT_TOKEN_REQUEST =
  '@farfetch/blackout-redux/REMOVE_PAYMENT_TOKEN_REQUEST';
/**
 * Action type dispatched when the remove payment token request succeeds.
 */
export const REMOVE_PAYMENT_TOKEN_SUCCESS =
  '@farfetch/blackout-redux/REMOVE_PAYMENT_TOKEN_SUCCESS';

/**
 * Action type dispatched when resetting the payment tokens.
 */
export const RESET_PAYMENT_TOKENS_STATE =
  '@farfetch/blackout-redux/RESET_PAYMENT_TOKENS_STATE';

/**
 * Action type dispatched when the fetch payment tokens request fails.
 */
export const FETCH_PAYMENT_TOKENS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_TOKENS_FAILURE';
/**
 * Action type dispatched when the fetch payment tokens request starts.
 */
export const FETCH_PAYMENT_TOKENS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_TOKENS_REQUEST';
/**
 * Action type dispatched when the fetch payment tokens request succeeds.
 */
export const FETCH_PAYMENT_TOKENS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_TOKENS_SUCCESS';

/**
 * Action type dispatched when the fetch payment intent charge request fails.
 */
export const FETCH_PAYMENT_INTENT_CHARGE_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_CHARGE_FAILURE';
/**
 * Action type dispatched when the fetch payment intent charge request starts.
 */
export const FETCH_PAYMENT_INTENT_CHARGE_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_CHARGE_REQUEST';
/**
 * Action type dispatched when the fetch payment intent charge request succeeds.
 */
export const FETCH_PAYMENT_INTENT_CHARGE_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_CHARGE_SUCCESS';

/**
 * Action type dispatched when resetting the payment intent charge.
 */
export const RESET_PAYMENT_INTENT_CHARGE_STATE =
  '@farfetch/blackout-redux/RESET_PAYMENT_INTENT_CHARGE_STATE';

/**
 * Action type dispatched when the fetch payment intent request fails.
 */
export const FETCH_PAYMENT_INTENT_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_FAILURE';
/**
 * Action type dispatched when the fetch payment intent request starts.
 */
export const FETCH_PAYMENT_INTENT_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_REQUEST';
/**
 * Action type dispatched when the fetch payment intent request succeeds.
 */
export const FETCH_PAYMENT_INTENT_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_INTENT_SUCCESS';

/**
 * Action type dispatched when the fetch payment methods request fails.
 */
export const FETCH_PAYMENT_METHODS_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_FAILURE';
/**
 * Action type dispatched when the fetch payment methods request starts.
 */
export const FETCH_PAYMENT_METHODS_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_REQUEST';
/**
 * Action type dispatched when the fetch payment methods request succeeds.
 */
export const FETCH_PAYMENT_METHODS_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_SUCCESS';

/**
 * Action type dispatched when the fetch payment methods by country and currency
 * request fails.
 */
export const FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_FAILURE';
/**
 * Action type dispatched when the fetch payment methods by country and currency
 * request starts.
 */
export const FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_REQUEST';
/**
 * Action type dispatched when the fetch payment methods by country and currency
 * request succeeds.
 */
export const FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_SUCCESS';

/**
 * Action type dispatched when the fetch payment methods by intent fails.
 */
export const FETCH_PAYMENT_METHODS_BY_INTENT_FAILURE =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_INTENT_FAILURE';
/**
 * Action type dispatched when the get payment methods by intent starts.
 */
export const FETCH_PAYMENT_METHODS_BY_INTENT_REQUEST =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_INTENT_REQUEST';
/**
 * Action type dispatched when the get payment methods by intent succeeds.
 */
export const FETCH_PAYMENT_METHODS_BY_INTENT_SUCCESS =
  '@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_INTENT_SUCCESS';

/**
 * Action type dispatched when the fetch user credit balance request fails.
 */
export const FETCH_USER_CREDIT_BALANCE_FAILURE =
  '@farfetch/blackout-redux/FETCH_USER_CREDIT_BALANCE_FAILURE';
/**
 * Action type dispatched when the fetch user credit balance request starts.
 */
export const FETCH_USER_CREDIT_BALANCE_REQUEST =
  '@farfetch/blackout-redux/FETCH_USER_CREDIT_BALANCE_REQUEST';
/**
 * Action type dispatched when the fetch user credit balance request succeeds.
 */
export const FETCH_USER_CREDIT_BALANCE_SUCCESS =
  '@farfetch/blackout-redux/FETCH_USER_CREDIT_BALANCE_SUCCESS';

/**
 * Action type dispatched when the fetch gift card balance request fails.
 */
export const FETCH_GIFT_CARD_BALANCE_FAILURE =
  '@farfetch/blackout-redux/FETCH_GIFT_CARD_BALANCE_FAILURE';
/**
 * Action type dispatched when the fetch gift card balance request starts.
 */
export const FETCH_GIFT_CARD_BALANCE_REQUEST =
  '@farfetch/blackout-redux/FETCH_GIFT_CARD_BALANCE_REQUEST';
/**
 * Action type dispatched when the fetch gift card balance request succeeds.
 */
export const FETCH_GIFT_CARD_BALANCE_SUCCESS =
  '@farfetch/blackout-redux/FETCH_GIFT_CARD_BALANCE_SUCCESS';

/**
 * Action type dispatched when the create payment intent charge request fails.
 */
export const CREATE_PAYMENT_INTENT_CHARGE_FAILURE =
  '@farfetch/blackout-redux/CREATE_PAYMENT_INTENT_CHARGE_FAILURE';
/**
 * Action type dispatched when the create payment intent charge request starts.
 */
export const CREATE_PAYMENT_INTENT_CHARGE_REQUEST =
  '@farfetch/blackout-redux/CREATE_PAYMENT_INTENT_CHARGE_REQUEST';
/**
 * Action type dispatched when the create payment intent charge request succeeds.
 */
export const CREATE_PAYMENT_INTENT_CHARGE_SUCCESS =
  '@farfetch/blackout-redux/CREATE_PAYMENT_INTENT_CHARGE_SUCCESS';

/**
 * Action type dispatched when resetting the payments state.
 */
export const RESET_PAYMENTS = '@farfetch/blackout-redux/RESET_PAYMENTS';
