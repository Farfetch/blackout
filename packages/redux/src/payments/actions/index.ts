/**
 * Payments actions.
 */
export { default as createPaymentIntentCharge } from './createPaymentIntentCharge.js';
export { default as createPaymentIntentInstrument } from './createPaymentIntentInstrument.js';
export { default as fetchUserCreditBalance } from './fetchUserCreditBalance.js';
export { default as fetchGiftCardBalance } from './fetchGiftCardBalance.js';
export { default as fetchPaymentIntentInstrument } from './fetchPaymentIntentInstrument.js';
export { default as fetchPaymentIntentInstruments } from './fetchPaymentIntentInstruments.js';
export { default as fetchPaymentIntent } from './fetchPaymentIntent.js';
export { default as fetchPaymentIntentCharge } from './fetchPaymentIntentCharge.js';
export { default as fetchPaymentMethodsByIntent } from './fetchPaymentMethodsByIntent.js';
export { default as fetchPaymentMethodsByCountryAndCurrency } from './fetchPaymentMethodsByCountryAndCurrency.js';
export { default as fetchPaymentTokens } from './fetchPaymentTokens.js';
export { default as removePaymentIntentInstrument } from './removePaymentIntentInstrument.js';
export { default as removePaymentToken } from './removePaymentToken.js';
export { default as resetGiftCardBalance } from './resetGiftCardBalance.js';
export { default as resetPaymentIntentInstruments } from './resetPaymentIntentInstruments.js';
export { default as resetPaymentIntentCharge } from './resetPaymentIntentCharge.js';
export { default as resetPaymentIntent } from './resetPaymentIntent.js';
export { default as resetPaymentMethods } from './resetPaymentMethods.js';
export { default as resetPayments } from './resetPayments.js';
export { default as resetPaymentTokens } from './resetPaymentTokens.js';
export { default as resetUserCreditBalance } from './resetUserCreditBalance.js';
export { default as updatePaymentIntentInstrument } from './updatePaymentIntentInstrument.js';
