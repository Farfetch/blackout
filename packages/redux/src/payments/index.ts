export * as paymentsActionTypes from './actionTypes.js';

export * from './actions/index.js';
export * from './actions/factories/index.js';
export * from './selectors.js';

export {
  default as paymentsReducer,
  entitiesMapper as paymentsEntitiesMapper,
} from './reducer.js';

export * from './types/index.js';
