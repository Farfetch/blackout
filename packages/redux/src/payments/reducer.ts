/**
 * @module payments/reducer
 * @category Payments
 * @subcategory Reducer
 */
import * as actionTypes from './actionTypes';
import { combineReducers } from 'redux';
import { createReducerWithResult } from '../helpers';
import { LOGOUT_SUCCESS } from '../authentication/actionTypes';
import get from 'lodash/get';
import omit from 'lodash/omit';
import type * as T from './types';
import type {
  Charges,
  Instrument,
  PaymentToken,
} from '@farfetch/blackout-client/payments/types';
import type { LogoutSuccessAction } from '../authentication/types';
import type {
  ReducerSwitch,
  StateWithResult,
  StateWithResultArray,
  StoreState,
} from '../types';

export const INITIAL_STATE: T.State = {
  charges: {
    error: null,
    isLoading: false,
    result: null,
  },
  creditBalance: {
    error: null,
    isLoading: false,
    result: null,
  },
  giftCardBalance: {
    error: null,
    isLoading: false,
    result: null,
  },
  instruments: {
    error: null,
    isLoading: false,
    result: null,
  },
  intent: {
    error: null,
    isLoading: false,
    result: null,
  },
  paymentMethods: {
    error: null,
    isLoading: false,
    result: null,
  },
  tokens: {
    error: null,
    isLoading: false,
    result: null,
  },
};

const charges = (
  state = INITIAL_STATE.charges,
  action:
    | T.ChargesFailureAction
    | T.ChargesRequestAction
    | T.ChargesSuccessAction
    | T.FetchChargesFailureAction
    | T.FetchChargesRequestAction
    | T.FetchChargesSuccessAction
    | T.ResetChargesSuccessAction
    | LogoutSuccessAction,
): StateWithResult<Charges> => {
  switch (action.type) {
    case actionTypes.CHARGE_REQUEST:
    case actionTypes.FETCH_CHARGES_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.charges.error,
        isLoading: true,
      };
    case actionTypes.CHARGE_FAILURE:
    case actionTypes.FETCH_CHARGES_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.CHARGE_SUCCESS:
    case actionTypes.FETCH_CHARGES_SUCCESS:
      return {
        error: INITIAL_STATE.charges.error,
        isLoading: false,
        result: {
          ...action.payload,
          chargeId: get(action.meta, 'chargeId'),
        },
      };
    case actionTypes.RESET_CHARGES_STATE:
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.charges;
    default:
      return state;
  }
};

const creditBalance = createReducerWithResult(
  'FETCH_CREDIT_BALANCE',
  INITIAL_STATE.creditBalance,
  actionTypes,
);

const giftCardBalance = createReducerWithResult(
  'FETCH_GIFT_CARD_BALANCE',
  INITIAL_STATE.giftCardBalance,
  actionTypes,
);

const instruments = (
  state = INITIAL_STATE.instruments,
  action:
    | T.CreateInstrumentsFailureAction
    | T.CreateInstrumentsRequestAction
    | T.CreateInstrumentsSuccessAction
    | T.FetchInstrumentFailureAction
    | T.FetchInstrumentRequestAction
    | T.FetchInstrumentSuccessAction
    | T.FetchInstrumentsFailureAction
    | T.FetchInstrumentsRequestAction
    | T.FetchInstrumentsSuccessAction
    | T.RemoveInstrumentFailureAction
    | T.RemoveInstrumentRequestAction
    | T.RemoveInstrumentSuccessAction
    | T.UpdateInstrumentFailureAction
    | T.UpdateInstrumentRequestAction
    | T.UpdateInstrumentSuccessAction
    | T.ResetInstrumentsSuccessAction
    | LogoutSuccessAction,
): StateWithResultArray<Instrument['id']> => {
  switch (action.type) {
    case actionTypes.FETCH_INSTRUMENT_REQUEST:
    case actionTypes.FETCH_INSTRUMENTS_REQUEST:
    case actionTypes.CREATE_INSTRUMENT_REQUEST:
    case actionTypes.UPDATE_INSTRUMENT_REQUEST:
    case actionTypes.REMOVE_INSTRUMENT_REQUEST:
      return {
        ...state,
        error: INITIAL_STATE.instruments.error,
        isLoading: true,
      };
    case actionTypes.FETCH_INSTRUMENT_FAILURE:
    case actionTypes.FETCH_INSTRUMENTS_FAILURE:
    case actionTypes.CREATE_INSTRUMENT_FAILURE:
    case actionTypes.UPDATE_INSTRUMENT_FAILURE:
    case actionTypes.REMOVE_INSTRUMENT_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };
    case actionTypes.FETCH_INSTRUMENT_SUCCESS:
    case actionTypes.FETCH_INSTRUMENTS_SUCCESS:
      return {
        error: INITIAL_STATE.instruments.error,
        isLoading: false,
        result: action.payload.result,
      };
    case actionTypes.CREATE_INSTRUMENT_SUCCESS:
    case actionTypes.UPDATE_INSTRUMENT_SUCCESS:
      return {
        ...state,
        error: INITIAL_STATE.instruments.error,
        isLoading: false,
      };
    case actionTypes.REMOVE_INSTRUMENT_SUCCESS:
      return {
        error: INITIAL_STATE.tokens.error,
        isLoading: false,
        result: state?.result?.filter(
          (instrumentId: Instrument['id']) =>
            instrumentId !== action.meta.instrumentId,
        ),
      };
    case actionTypes.RESET_INSTRUMENTS_STATE:
    case LOGOUT_SUCCESS:
      return INITIAL_STATE.instruments;
    default:
      return state;
  }
};

const intent = createReducerWithResult(
  'FETCH_INTENT',
  INITIAL_STATE.intent,
  actionTypes,
);

const paymentMethods = createReducerWithResult(
  [
    'FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY',
    'FETCH_PAYMENT_METHODS_BY_INTENT',
  ],
  INITIAL_STATE.paymentMethods,
  actionTypes,
  false,
  true,
);

const tokens = (
  state = INITIAL_STATE.tokens,
  action:
    | T.FetchPaymentTokensFailureAction
    | T.FetchPaymentTokensRequestAction
    | T.FetchPaymentTokensSuccessAction
    | T.RemovePaymentTokensFailureAction
    | T.RemovePaymentTokensRequestAction
    | T.RemovePaymentTokensSuccessAction
    | LogoutSuccessAction,
): StateWithResultArray<PaymentToken['id']> => {
  switch (action.type) {
    case actionTypes.FETCH_PAYMENT_TOKENS_REQUEST:
    case actionTypes.REMOVE_PAYMENT_TOKEN_REQUEST:
      return {
        ...state,
        isLoading: true,
        error: INITIAL_STATE.tokens.error,
      };

    case actionTypes.FETCH_PAYMENT_TOKENS_SUCCESS:
      return {
        ...state,
        result: action.payload.result,
        isLoading: false,
      };

    case actionTypes.REMOVE_PAYMENT_TOKEN_SUCCESS:
      return {
        ...state,
        result: state?.result?.filter(
          (tokenId: PaymentToken['id']) => tokenId !== action.meta.id,
        ),
        isLoading: false,
      };

    case actionTypes.FETCH_PAYMENT_TOKENS_FAILURE:
    case actionTypes.REMOVE_PAYMENT_TOKEN_FAILURE:
      return {
        ...state,
        error: action.payload.error,
        isLoading: false,
      };

    case LOGOUT_SUCCESS:
      return INITIAL_STATE.tokens;
    default:
      return state;
  }
};

export const entitiesMapper = {
  [actionTypes.REMOVE_PAYMENT_TOKEN_SUCCESS as typeof actionTypes.REMOVE_PAYMENT_TOKEN_SUCCESS]:
    (
      state: StoreState['entities'],
      action: T.RemovePaymentTokensSuccessAction,
    ): StoreState['entities'] => {
      const { id } = action.meta;
      const currentPaymentTokens = state.paymentTokens;

      return {
        ...state,
        paymentTokens: omit(currentPaymentTokens, id),
      };
    },
  [actionTypes.REMOVE_INSTRUMENT_SUCCESS as typeof actionTypes.REMOVE_INSTRUMENT_SUCCESS]:
    (
      state: StoreState['entities'],
      action: T.RemoveInstrumentSuccessAction,
    ): StoreState['entities'] => {
      const { instrumentId } = action.meta;
      const currentInstruments = state.instruments;

      return {
        ...state,
        instruments: omit(currentInstruments, instrumentId),
      };
    },
  [actionTypes.RESET_INSTRUMENTS_STATE as typeof actionTypes.RESET_INSTRUMENTS_STATE]:
    (state: StoreState['entities']): StoreState['entities'] => {
      return {
        ...state,
        instruments: {},
      };
    },
  [LOGOUT_SUCCESS as typeof LOGOUT_SUCCESS]: (
    state: StoreState['entities'],
  ): StoreState['entities'] => {
    return {
      ...state,
      instruments: {},
      paymentTokens: {},
      checkoutOrders: {},
    };
  },
};

export const getTokens = (state: T.State): T.State['tokens'] => state.tokens;
export const getInstruments = (state: T.State): T.State['instruments'] =>
  state.instruments;
export const getGiftCardBalance = (
  state: T.State,
): T.State['giftCardBalance'] => state.giftCardBalance;
export const getCreditBalance = (state: T.State): T.State['creditBalance'] =>
  state.creditBalance;
export const getIntent = (state: T.State): T.State['intent'] => state.intent;
export const getCharges = (state: T.State): T.State['charges'] => state.charges;
export const getPaymentMethods = (state: T.State): T.State['paymentMethods'] =>
  state.paymentMethods;

const reducers = combineReducers({
  charges,
  creditBalance,
  giftCardBalance,
  instruments,
  intent,
  paymentMethods,
  tokens,
});

/**
 * Reducer for payments state.
 *
 * @function paymentsReducer
 * @static
 *
 * @param {object} state - Current redux state.
 * @param {object} action - Action dispatched.
 *
 * @returns {object} New state.
 */
const paymentsReducer: ReducerSwitch<
  T.State,
  | T.ChargesAction
  | T.CreateInstrumentsAction
  | T.FetchCreditBalanceAction
  | T.FetchGiftCardBalanceAction
  | T.FetchChargesAction
  | T.FetchInstrumentAction
  | T.FetchInstrumentsAction
  | T.FetchIntentAction
  | T.FetchPaymentTokensAction
  | T.FetchPaymentMethodsAction
  | T.FetchPaymentMethodsByCountryAndCurrencyAction
  | T.FetchPaymentMethodsByIntentAction
  | T.RemoveInstrumentAction
  | T.RemovePaymentTokensAction
  | T.UpdateInstrumentAction
  | T.ResetChargesAction
  | T.ResetInstrumentsAction
> = (state, action) => reducers(state, action);

export default paymentsReducer;
