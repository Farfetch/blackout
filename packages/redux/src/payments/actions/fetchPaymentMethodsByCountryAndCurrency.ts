import { fetchPaymentMethodsByCountryAndCurrencyFactory } from './factories';
import { getPaymentMethodsByCountryAndCurrency } from '@farfetch/blackout-client/payments';

/**
 * Fetch payment methods by country and currency.
 *
 * @memberof module:payments/actions
 *
 * @name fetchPaymentMethodsByCountryAndCurrency
 *
 * @type {FetchPaymentMethodsByCountryAndCurrencyThunkFactory}
 */
export default fetchPaymentMethodsByCountryAndCurrencyFactory(
  getPaymentMethodsByCountryAndCurrency,
);
