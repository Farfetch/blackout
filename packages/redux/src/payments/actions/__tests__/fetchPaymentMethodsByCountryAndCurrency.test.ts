import * as actionTypes from '../../actionTypes';
import { getPaymentMethodsByCountryAndCurrency } from '@farfetch/blackout-client';
import { INITIAL_STATE } from '../../reducer';
import { mockmockFetchPaymentMethodsResponse } from 'tests/__fixtures__/payments';
import { mockStore } from '../../../../tests';
import fetchPaymentMethodsByCountryAndCurrency from '../fetchPaymentMethodsByCountryAndCurrency';
import find from 'lodash/find';

jest.mock('@farfetch/blackout-client', () => ({
  ...jest.requireActual('@farfetch/blackout-client'),
  getPaymentMethodsByCountryAndCurrency: jest.fn(),
}));

const paymentsMockStore = (state = {}) =>
  mockStore({ payments: INITIAL_STATE }, state);

const expectedConfig = undefined;
let store;

describe('fetchPaymentMethodsByCountryAndCurrency() action creator', () => {
  beforeEach(() => {
    jest.clearAllMocks();
    store = paymentsMockStore();
  });

  it('should create the correct actions when the fetch payment methods with country and currency procedure fails', async () => {
    const expectedError = new Error(
      'fetch payment methods with country and currency error',
    );

    getPaymentMethodsByCountryAndCurrency.mockRejectedValueOnce(expectedError);
    expect.assertions(4);

    try {
      await store.dispatch(fetchPaymentMethodsByCountryAndCurrency());
    } catch (error) {
      expect(error).toBe(expectedError);
      expect(getPaymentMethodsByCountryAndCurrency).toHaveBeenCalledTimes(1);
      expect(getPaymentMethodsByCountryAndCurrency).toHaveBeenCalledWith(
        expectedConfig,
      );
      expect(store.getActions()).toEqual(
        expect.arrayContaining([
          {
            type: actionTypes.FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_REQUEST,
          },
          {
            type: actionTypes.FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_FAILURE,

            payload: { error: expectedError },
          },
        ]),
      );
    }
  });

  it('should create the correct actions when the fetch payment methods with country and currency procedure is successful', async () => {
    getPaymentMethodsByCountryAndCurrency.mockResolvedValueOnce(
      mockmockFetchPaymentMethodsResponse,
    );
    await store.dispatch(fetchPaymentMethodsByCountryAndCurrency());

    const actionResults = store.getActions();

    expect(getPaymentMethodsByCountryAndCurrency).toHaveBeenCalledTimes(1);
    expect(getPaymentMethodsByCountryAndCurrency).toHaveBeenCalledWith(
      expectedConfig,
    );
    expect(actionResults).toMatchObject([
      {
        type: actionTypes.FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_REQUEST,
      },
      {
        type: actionTypes.FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_SUCCESS,
        payload: mockmockFetchPaymentMethodsResponse,
      },
    ]);
    expect(
      find(actionResults, {
        type: actionTypes.FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_SUCCESS,
      }),
    ).toMatchSnapshot(
      'fetch payment methods with country and currency success payload',
    );
  });
});
