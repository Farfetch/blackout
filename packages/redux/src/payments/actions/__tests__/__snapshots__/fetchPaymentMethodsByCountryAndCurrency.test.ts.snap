// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchPaymentMethodsByCountryAndCurrency() action creator should create the correct actions when the fetch payment methods with country and currency procedure is successful: fetch payment methods with country and currency success payload 1`] = `
Object {
  "payload": undefined,
  "type": "@farfetch/blackout-redux/FETCH_PAYMENT_METHODS_BY_COUNTRY_AND_CURRENCY_SUCCESS",
}
`;
